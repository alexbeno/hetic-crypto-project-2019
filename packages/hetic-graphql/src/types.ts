import { context, dataSources } from './server'

export interface Crypto {
  id: number
  symbol: string
  name?: string
  starred: boolean
}

export interface Value {
  id: number
  cryptoId: number
  date: Date
  value: number
}

export type Context = typeof context & {
  dataSources: ReturnType<typeof dataSources>
}
