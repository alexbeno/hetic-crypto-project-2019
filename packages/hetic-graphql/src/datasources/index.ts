export * from './crypto'
export * from './values'
