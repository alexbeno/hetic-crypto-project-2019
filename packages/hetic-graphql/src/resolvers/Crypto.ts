import { CryptoResolvers } from '../generated/graphqlgen'

export const Crypto: CryptoResolvers.Type = {
  ...CryptoResolvers.defaultResolvers,
}
