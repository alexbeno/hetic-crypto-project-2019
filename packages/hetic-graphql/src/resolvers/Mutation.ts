import { MutationResolvers } from '../generated/graphqlgen'

export const Mutation: MutationResolvers.Type = {
  ...MutationResolvers.defaultResolvers,
  updateCrypto: (_parent, args, ctx) => ctx.dataSources.cryptoAPI.updateOne(args),
}
