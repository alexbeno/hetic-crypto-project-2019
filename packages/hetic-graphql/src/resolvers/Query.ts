import { QueryResolvers } from '../generated/graphqlgen'

export const Query: QueryResolvers.Type = {
  ...QueryResolvers.defaultResolvers,
  cryptoes: (_parent, args, ctx) => ctx.dataSources.cryptoAPI.findAll(args),
  values: (_parent, args, ctx) => ctx.dataSources.valuesAPI.findAll(args),
}
