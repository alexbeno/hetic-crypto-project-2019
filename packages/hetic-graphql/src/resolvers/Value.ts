import { ValueResolvers } from '../generated/graphqlgen'

export const Value: ValueResolvers.Type = {
  ...ValueResolvers.defaultResolvers,
  createdAt: parent => parent.date.toISOString(),
  crypto: (parent, _args, ctx) => ctx.dataSources.cryptoAPI.findOne(parent.cryptoId),
}
