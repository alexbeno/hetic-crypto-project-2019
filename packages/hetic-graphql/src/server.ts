import { ApolloServer, gql } from 'apollo-server'
import fs from 'fs'

// The GraphQL schema
const typeDefs = gql`
  ${fs.readFileSync(__dirname.concat('/schema.graphql'), 'utf8')}
`

// A map of functions which return data for the schema.
const resolvers = {
  Query: {
    hello: () => 'world',
  },
}

const server = new ApolloServer({
  typeDefs,
  resolvers,
})

server.listen().then(({ url }) => {
  console.log(`ğŸš€ Server ready at ${url}`)
})
