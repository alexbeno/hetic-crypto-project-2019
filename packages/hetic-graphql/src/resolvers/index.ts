import { Resolvers } from '../generated/graphqlgen'
import { Crypto } from './Crypto'
import { Mutation } from './Mutation'
import { Query } from './Query'
import { Value } from './Value'

// Resolvers define the technique for fetching the types in the schema.
const resolvers: Resolvers = {
  Query,
  Crypto,
  Mutation,
  Value,
}

export default resolvers
