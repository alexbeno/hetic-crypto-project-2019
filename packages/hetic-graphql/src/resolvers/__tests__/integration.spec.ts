import { ApolloServer, gql } from 'apollo-server'
import { createTestClient } from 'apollo-server-testing'

import resolvers from '../'
import { CryptoAPI, ValuesAPI } from '../../datasources'
import { typeDefs } from '../../server'
import { Crypto, Value } from '../../types'

jest.mock('../../datasources/crypto', () => {
  class CryptoAPI {
    findAll(): Crypto[] {
      return [
        {
          id: 1,
          symbol: 'BTC',
          name: 'Bitcoin',
          starred: false,
        },
      ]
    }
  }
  return { CryptoAPI }
})

jest.mock('../../datasources/values', () => {
  class ValuesAPI {
    findAll(): Value[] {
      return [
        {
          id: 1,
          cryptoId: 1,
          date: new Date('2019-04-21T22:00:00.000Z'),
          value: 100000.1,
        },
      ]
    }
  }
  return { ValuesAPI }
})

const GET_VALUES = gql`
  query {
    values {
      id
      createdAt
      value
    }
  }
`

const GET_CRYPTOES = gql`
  query {
    cryptoes {
      id
      symbol
      name
      starred
    }
  }
`

describe('Queries', () => {
  let client: ReturnType<typeof createTestClient>

  beforeAll(() => {
    const server = new ApolloServer({
      typeDefs,
      resolvers,
      dataSources: () => ({ cryptoAPI: new CryptoAPI(), valuesAPI: new ValuesAPI() }),
    })
    client = createTestClient(server)
  })

  it('fetches list of values', async () => {
    expect(await client.query({ query: GET_VALUES })).toMatchSnapshot()
  })

  it('fetches list of cryptocurrencies', async () => {
    expect(await client.query({ query: GET_CRYPTOES })).toMatchSnapshot()
  })
})
