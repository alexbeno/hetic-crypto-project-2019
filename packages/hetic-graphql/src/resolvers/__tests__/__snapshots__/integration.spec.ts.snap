// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries fetches list of cryptocurrencies 1`] = `
Object {
  "data": Object {
    "cryptoes": Array [
      Object {
        "id": 1,
        "name": "Bitcoin",
        "starred": false,
        "symbol": "BTC",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches list of values 1`] = `
Object {
  "data": Object {
    "values": Array [
      Object {
        "createdAt": "2019-04-21T22:00:00.000Z",
        "id": 1,
        "value": 100000.1,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
